controllers/cloudstackmachine_controller.go

```diff
@@ -268,6 +269,17 @@ func (r *CloudStackMachineReconciliationRunner) GetOrCreateMachineStateChecker()
 }
 
 func (r *CloudStackMachineReconciliationRunner) ReconcileDelete() (retRes ctrl.Result, reterr error) {
+	if r.ReconciliationSubject.Spec.InstanceID == nil {
+		// InstanceID is not set until deploying VM finishes which can take minutes, and CloudStack Machine can be deleted before VM deployment complete.
+		// ResolveVMInstanceDetails can get InstanceID by CS machine name
+		err := r.CSClient.ResolveVMInstanceDetails(r.ReconciliationSubject)
```

Comment:
does this method automatically set instanceID? Should it return the value, instead of just an error?

Reply:
Yes, ResolveVMInstanceDetails automatically set the instanceID. It is an existed function, we are trying to re-call it to get latest info from CloudStack management API. And instanceID is part of the thing it sets.

Reply:
ResolveVMInstanceDetails is one of 3 function exposed through VMIface

1. GetOrCreateVMInstance
2. ResolveVMInstanceDetails
3. DestroyVMInstance

Setting InstanceID is only part of its function, and change its func name to indicate that it sets instanceID kind of misleading since it also setting other things. 

I can add a comments before setting the value, so that future developer know its usage when refactoring code.
```
        csMachine.Spec.ProviderID = pointer.StringPtr(fmt.Sprintf("cloudstack:///%s", vmResponse.Id))
	// InstanceID is later used as required parameter to destroy VM.
	csMachine.Spec.InstanceID = pointer.StringPtr(vmResponse.Id)
	
```